### ISSUE TITLE
[pango:x86-windows-static] build failure

### ISSUE THREAD

**Issue by bby-co** [microsoft/vcpkg#7645]
EDIT: Issue is fixed. Read my last post for further information.

Building package pango[core]:x86-windows-static...
-- Using cached C:/Program Files (x86)/vcpkg/downloads/pango-1.40.11.tar.xz
-- Using source at C:/Program Files (x86)/vcpkg/buildtrees/pango/src/1.40.11-f997ae867d
-- Configuring x86-windows-static
CMake Error at scripts/cmake/vcpkg_execute_required_process.cmake:58 (message):
    Command failed: ninja -v
    Working Directory: C:/Program Files (x86)/vcpkg/buildtrees/pango/x86-windows-static-rel/vcpkg-parallel-configure
    Error code: 1
    See logs for more information:
      C:\Program Files (x86)\vcpkg\buildtrees\pango\config-x86-windows-static-out.log

Call Stack (most recent call first):
  scripts/cmake/vcpkg_configure_cmake.cmake:290 (vcpkg_execute_required_process)
  ports/pango/portfile.cmake:18 (vcpkg_configure_cmake)
  scripts/ports.cmake:74 (include)

Error: Building package pango:x86-windows-static failed with: BUILD_FAILED
Please ensure you're using the latest portfiles with `.\vcpkg update`, then
submit an issue at https://github.com/Microsoft/vcpkg/issues including:
  Package: pango:x86-windows-static
  Vcpkg version: 2019.07.18-nohash

**Comment by bby-co** [comment]
The issue is resolved by running `vcpkg install harfbuzz:x86-windows-static`, however now encountering a different problem.

```
-- Using source at C:/Program Files (x86)/vcpkg/buildtrees/pango/src/1.40.11-f997ae867d
-- Configuring x86-windows-static
-- Building x86-windows-static-dbg
CMake Error at scripts/cmake/vcpkg_execute_build_process.cmake:136 (message):
    Command failed: "C:/Program Files (x86)/vcpkg/downloads/tools/cmake-3.14.0-windows/cmake-3.14.0-win32-x86/bin/cmake.exe" --build . --config Debug --target install -- -v
    Working Directory: C:/Program Files (x86)/vcpkg/buildtrees/pango/x86-windows-static-dbg
    See logs for more information:
      C:\Program Files (x86)\vcpkg\buildtrees\pango\install-x86-windows-static-dbg-out.log

Call Stack (most recent call first):
  scripts/cmake/vcpkg_build_cmake.cmake:96 (vcpkg_execute_build_process)
  scripts/cmake/vcpkg_install_cmake.cmake:24 (vcpkg_build_cmake)
  ports/pango/portfile.cmake:25 (vcpkg_install_cmake)
  scripts/ports.cmake:74 (include)


Error: Building package pango:x86-windows-static failed with: BUILD_FAILED
Please ensure you're using the latest portfiles with `.\vcpkg update`, then
submit an issue at https://github.com/Microsoft/vcpkg/issues including:
  Package: pango:x86-windows-static
  Vcpkg version: 2019.07.18-nohash
```

Snip from install-x86-windows-static-dbg-out.log:
```
[53/61] C:\PROGRA~2\MICROS~2\2019\COMMUN~1\VC\Tools\MSVC\1422~1.279\bin\Hostx64\x86\cl.exe   -DG_DISABLE_SINGLE_INCLUDES -DG_LOG_DOMAIN=\"Pango\" -DHAVE_CAIRO_FREETYPE=1 -DHAVE_CONFIG_H -DLIBDIR=\"/dummy/lib\" -DPANGO_ENABLE_BACKEND -DPANGO_ENABLE_ENGINE -DSYSCONFDIR=\"/dummy/etc\" -I"C:\Program Files (x86)\vcpkg\buildtrees\pango\src\1.40.11-f997ae867d\." -I"C:\Program Files (x86)\vcpkg\buildtrees\pango\src\1.40.11-f997ae867d\.\pango" -I. -I"C:\Program Files (x86)\vcpkg\installed\x86-windows-static\include" -I"C:\Program Files (x86)\vcpkg\installed\x86-windows-static\include\harfbuzz" /DWIN32 /D_WINDOWS /W3 /utf-8 /MP  /utf-8 /D_DEBUG /MTd /Z7 /Ob0 /Od /RTC1 /showIncludes /FoCMakeFiles\pangoft2.dir\pango\pangoft2-fontmap.c.obj /FdCMakeFiles\pangoft2.dir\pangoft2.pdb /FS -c "C:\Program Files (x86)\vcpkg\buildtrees\pango\src\1.40.11-f997ae867d\pango\pangoft2-fontmap.c"
Microsoft (R) C/C++ Optimizing Compiler Version 19.22.27905 for x86
Copyright (C) Microsoft Corporation.  All rights reserved.

[54/61] C:\PROGRA~2\MICROS~2\2019\COMMUN~1\VC\Tools\MSVC\1422~1.279\bin\Hostx64\x86\cl.exe   -DG_DISABLE_SINGLE_INCLUDES -DG_LOG_DOMAIN=\"Pango\" -DHAVE_CAIRO_FREETYPE=1 -DHAVE_CONFIG_H -DLIBDIR=\"/dummy/lib\" -DPANGO_ENABLE_BACKEND -DPANGO_ENABLE_ENGINE -DSYSCONFDIR=\"/dummy/etc\" -I"C:\Program Files (x86)\vcpkg\buildtrees\pango\src\1.40.11-f997ae867d\." -I"C:\Program Files (x86)\vcpkg\buildtrees\pango\src\1.40.11-f997ae867d\.\pango" -I. -I"C:\Program Files (x86)\vcpkg\installed\x86-windows-static\include" -I"C:\Program Files (x86)\vcpkg\installed\x86-windows-static\include\harfbuzz" /DWIN32 /D_WINDOWS /W3 /utf-8 /MP  /utf-8 /D_DEBUG /MTd /Z7 /Ob0 /Od /RTC1 /showIncludes /FoCMakeFiles\pangoft2.dir\pango\pango-ot-info.c.obj /FdCMakeFiles\pangoft2.dir\pangoft2.pdb /FS -c "C:\Program Files (x86)\vcpkg\buildtrees\pango\src\1.40.11-f997ae867d\pango\pango-ot-info.c"
FAILED: CMakeFiles/pangoft2.dir/pango/pango-ot-info.c.obj 
C:\PROGRA~2\MICROS~2\2019\COMMUN~1\VC\Tools\MSVC\1422~1.279\bin\Hostx64\x86\cl.exe   -DG_DISABLE_SINGLE_INCLUDES -DG_LOG_DOMAIN=\"Pango\" -DHAVE_CAIRO_FREETYPE=1 -DHAVE_CONFIG_H -DLIBDIR=\"/dummy/lib\" -DPANGO_ENABLE_BACKEND -DPANGO_ENABLE_ENGINE -DSYSCONFDIR=\"/dummy/etc\" -I"C:\Program Files (x86)\vcpkg\buildtrees\pango\src\1.40.11-f997ae867d\." -I"C:\Program Files (x86)\vcpkg\buildtrees\pango\src\1.40.11-f997ae867d\.\pango" -I. -I"C:\Program Files (x86)\vcpkg\installed\x86-windows-static\include" -I"C:\Program Files (x86)\vcpkg\installed\x86-windows-static\include\harfbuzz" /DWIN32 /D_WINDOWS /W3 /utf-8 /MP  /utf-8 /D_DEBUG /MTd /Z7 /Ob0 /Od /RTC1 /showIncludes /FoCMakeFiles\pangoft2.dir\pango\pango-ot-info.c.obj /FdCMakeFiles\pangoft2.dir\pangoft2.pdb /FS -c "C:\Program Files (x86)\vcpkg\buildtrees\pango\src\1.40.11-f997ae867d\pango\pango-ot-info.c"
Microsoft (R) C/C++ Optimizing Compiler Version 19.22.27905 for x86
Copyright (C) Microsoft Corporation.  All rights reserved.

C:\Program Files (x86)\vcpkg\buildtrees\pango\src\1.40.11-f997ae867d\pango\pango-ot-private.h(30): fatal error C1083: Cannot open include file: 'hb-glib.h': No such file or directory
[55/61] C:\PROGRA~2\MICROS~2\2019\COMMUN~1\VC\Tools\MSVC\1422~1.279\bin\Hostx64\x86\cl.exe   -DG_DISABLE_SINGLE_INCLUDES -DG_LOG_DOMAIN=\"Pango\" -DHAVE_CAIRO_FREETYPE=1 -DHAVE_CONFIG_H -DLIBDIR=\"/dummy/lib\" -DPANGO_ENABLE_BACKEND -DPANGO_ENABLE_ENGINE -DSYSCONFDIR=\"/dummy/etc\" -I"C:\Program Files (x86)\vcpkg\buildtrees\pango\src\1.40.11-f997ae867d\." -I"C:\Program Files (x86)\vcpkg\buildtrees\pango\src\1.40.11-f997ae867d\.\pango" -I. -I"C:\Program Files (x86)\vcpkg\installed\x86-windows-static\include" -I"C:\Program Files (x86)\vcpkg\installed\x86-windows-static\include\harfbuzz" /DWIN32 /D_WINDOWS /W3 /utf-8 /MP  /utf-8 /D_DEBUG /MTd /Z7 /Ob0 /Od /RTC1 /showIncludes /FoCMakeFiles\pangoft2.dir\pango\pango-ot-ruleset.c.obj /FdCMakeFiles\pangoft2.dir\pangoft2.pdb /FS -c "C:\Program Files (x86)\vcpkg\buildtrees\pango\src\1.40.11-f997ae867d\pango\pango-ot-ruleset.c"
FAILED: CMakeFiles/pangoft2.dir/pango/pango-ot-ruleset.c.obj 
C:\PROGRA~2\MICROS~2\2019\COMMUN~1\VC\Tools\MSVC\1422~1.279\bin\Hostx64\x86\cl.exe   -DG_DISABLE_SINGLE_INCLUDES -DG_LOG_DOMAIN=\"Pango\" -DHAVE_CAIRO_FREETYPE=1 -DHAVE_CONFIG_H -DLIBDIR=\"/dummy/lib\" -DPANGO_ENABLE_BACKEND -DPANGO_ENABLE_ENGINE -DSYSCONFDIR=\"/dummy/etc\" -I"C:\Program Files (x86)\vcpkg\buildtrees\pango\src\1.40.11-f997ae867d\." -I"C:\Program Files (x86)\vcpkg\buildtrees\pango\src\1.40.11-f997ae867d\.\pango" -I. -I"C:\Program Files (x86)\vcpkg\installed\x86-windows-static\include" -I"C:\Program Files (x86)\vcpkg\installed\x86-windows-static\include\harfbuzz" /DWIN32 /D_WINDOWS /W3 /utf-8 /MP  /utf-8 /D_DEBUG /MTd /Z7 /Ob0 /Od /RTC1 /showIncludes /FoCMakeFiles\pangoft2.dir\pango\pango-ot-ruleset.c.obj /FdCMakeFiles\pangoft2.dir\pangoft2.pdb /FS -c "C:\Program Files (x86)\vcpkg\buildtrees\pango\src\1.40.11-f997ae867d\pango\pango-ot-ruleset.c"
Microsoft (R) C/C++ Optimizing Compiler Version 19.22.27905 for x86
Copyright (C) Microsoft Corporation.  All rights reserved.

C:\Program Files (x86)\vcpkg\buildtrees\pango\src\1.40.11-f997ae867d\pango\pango-ot-private.h(30): fatal error C1083: Cannot open include file: 'hb-glib.h': No such file or directory
[56/61] C:\PROGRA~2\MICROS~2\2019\COMMUN~1\VC\Tools\MSVC\1422~1.279\bin\Hostx64\x86\cl.exe   -DG_DISABLE_SINGLE_INCLUDES -DG_LOG_DOMAIN=\"Pango\" -DHAVE_CAIRO_FREETYPE=1 -DHAVE_CONFIG_H -DLIBDIR=\"/dummy/lib\" -DPANGO_ENABLE_BACKEND -DPANGO_ENABLE_ENGINE -DSYSCONFDIR=\"/dummy/etc\" -I"C:\Program Files (x86)\vcpkg\buildtrees\pango\src\1.40.11-f997ae867d\." -I"C:\Program Files (x86)\vcpkg\buildtrees\pango\src\1.40.11-f997ae867d\.\pango" -I. -I"C:\Program Files (x86)\vcpkg\installed\x86-windows-static\include" -I"C:\Program Files (x86)\vcpkg\installed\x86-windows-static\include\harfbuzz" /DWIN32 /D_WINDOWS /W3 /utf-8 /MP  /utf-8 /D_DEBUG /MTd /Z7 /Ob0 /Od /RTC1 /showIncludes /FoCMakeFiles\pangoft2.dir\pango\pango-ot-buffer.c.obj /FdCMakeFiles\pangoft2.dir\pangoft2.pdb /FS -c "C:\Program Files (x86)\vcpkg\buildtrees\pango\src\1.40.11-f997ae867d\pango\pango-ot-buffer.c"
FAILED: CMakeFiles/pangoft2.dir/pango/pango-ot-buffer.c.obj 
C:\PROGRA~2\MICROS~2\2019\COMMUN~1\VC\Tools\MSVC\1422~1.279\bin\Hostx64\x86\cl.exe   -DG_DISABLE_SINGLE_INCLUDES -DG_LOG_DOMAIN=\"Pango\" -DHAVE_CAIRO_FREETYPE=1 -DHAVE_CONFIG_H -DLIBDIR=\"/dummy/lib\" -DPANGO_ENABLE_BACKEND -DPANGO_ENABLE_ENGINE -DSYSCONFDIR=\"/dummy/etc\" -I"C:\Program Files (x86)\vcpkg\buildtrees\pango\src\1.40.11-f997ae867d\." -I"C:\Program Files (x86)\vcpkg\buildtrees\pango\src\1.40.11-f997ae867d\.\pango" -I. -I"C:\Program Files (x86)\vcpkg\installed\x86-windows-static\include" -I"C:\Program Files (x86)\vcpkg\installed\x86-windows-static\include\harfbuzz" /DWIN32 /D_WINDOWS /W3 /utf-8 /MP  /utf-8 /D_DEBUG /MTd /Z7 /Ob0 /Od /RTC1 /showIncludes /FoCMakeFiles\pangoft2.dir\pango\pango-ot-buffer.c.obj /FdCMakeFiles\pangoft2.dir\pangoft2.pdb /FS -c "C:\Program Files (x86)\vcpkg\buildtrees\pango\src\1.40.11-f997ae867d\pango\pango-ot-buffer.c"
Microsoft (R) C/C++ Optimizing Compiler Version 19.22.27905 for x86
Copyright (C) Microsoft Corporation.  All rights reserved.

C:\Program Files (x86)\vcpkg\buildtrees\pango\src\1.40.11-f997ae867d\pango\pango-ot-private.h(30): fatal error C1083: Cannot open include file: 'hb-glib.h': No such file or directory
[57/61] C:\PROGRA~2\MICROS~2\2019\COMMUN~1\VC\Tools\MSVC\1422~1.279\bin\Hostx64\x86\cl.exe   -DG_DISABLE_SINGLE_INCLUDES -DG_LOG_DOMAIN=\"Pango\" -DHAVE_CAIRO_FREETYPE=1 -DHAVE_CONFIG_H -DLIBDIR=\"/dummy/lib\" -DPANGO_ENABLE_BACKEND -DPANGO_ENABLE_ENGINE -DSYSCONFDIR=\"/dummy/etc\" -I"C:\Program Files (x86)\vcpkg\buildtrees\pango\src\1.40.11-f997ae867d\." -I"C:\Program Files (x86)\vcpkg\buildtrees\pango\src\1.40.11-f997ae867d\.\pango" -I. -I"C:\Program Files (x86)\vcpkg\installed\x86-windows-static\include" -I"C:\Program Files (x86)\vcpkg\installed\x86-windows-static\include\harfbuzz" /DWIN32 /D_WINDOWS /W3 /utf-8 /MP  /utf-8 /D_DEBUG /MTd /Z7 /Ob0 /Od /RTC1 /showIncludes /FoCMakeFiles\pangoft2.dir\pango\pangofc-shape.c.obj /FdCMakeFiles\pangoft2.dir\pangoft2.pdb /FS -c "C:\Program Files (x86)\vcpkg\buildtrees\pango\src\1.40.11-f997ae867d\pango\pangofc-shape.c"
FAILED: CMakeFiles/pangoft2.dir/pango/pangofc-shape.c.obj 
C:\PROGRA~2\MICROS~2\2019\COMMUN~1\VC\Tools\MSVC\1422~1.279\bin\Hostx64\x86\cl.exe   -DG_DISABLE_SINGLE_INCLUDES -DG_LOG_DOMAIN=\"Pango\" -DHAVE_CAIRO_FREETYPE=1 -DHAVE_CONFIG_H -DLIBDIR=\"/dummy/lib\" -DPANGO_ENABLE_BACKEND -DPANGO_ENABLE_ENGINE -DSYSCONFDIR=\"/dummy/etc\" -I"C:\Program Files (x86)\vcpkg\buildtrees\pango\src\1.40.11-f997ae867d\." -I"C:\Program Files (x86)\vcpkg\buildtrees\pango\src\1.40.11-f997ae867d\.\pango" -I. -I"C:\Program Files (x86)\vcpkg\installed\x86-windows-static\include" -I"C:\Program Files (x86)\vcpkg\installed\x86-windows-static\include\harfbuzz" /DWIN32 /D_WINDOWS /W3 /utf-8 /MP  /utf-8 /D_DEBUG /MTd /Z7 /Ob0 /Od /RTC1 /showIncludes /FoCMakeFiles\pangoft2.dir\pango\pangofc-shape.c.obj /FdCMakeFiles\pangoft2.dir\pangoft2.pdb /FS -c "C:\Program Files (x86)\vcpkg\buildtrees\pango\src\1.40.11-f997ae867d\pango\pangofc-shape.c"
Microsoft (R) C/C++ Optimizing Compiler Version 19.22.27905 for x86
Copyright (C) Microsoft Corporation.  All rights reserved.

C:\Program Files (x86)\vcpkg\buildtrees\pango\src\1.40.11-f997ae867d\pango\pangofc-shape.c(31): fatal error C1083: Cannot open include file: 'hb-glib.h': No such file or directory
[58/61] C:\PROGRA~2\MICROS~2\2019\COMMUN~1\VC\Tools\MSVC\1422~1.279\bin\Hostx64\x86\cl.exe   -DG_DISABLE_SINGLE_INCLUDES -DG_LOG_DOMAIN=\"Pango\" -DHAVE_CAIRO_FREETYPE=1 -DHAVE_CONFIG_H -DLIBDIR=\"/dummy/lib\" -DPANGO_ENABLE_BACKEND -DPANGO_ENABLE_ENGINE -DSYSCONFDIR=\"/dummy/etc\" -I"C:\Program Files (x86)\vcpkg\buildtrees\pango\src\1.40.11-f997ae867d\." -I"C:\Program Files (x86)\vcpkg\buildtrees\pango\src\1.40.11-f997ae867d\.\pango" -I. -I"C:\Program Files (x86)\vcpkg\installed\x86-windows-static\include" -I"C:\Program Files (x86)\vcpkg\installed\x86-windows-static\include\harfbuzz" /DWIN32 /D_WINDOWS /W3 /utf-8 /MP  /utf-8 /D_DEBUG /MTd /Z7 /Ob0 /Od /RTC1 /showIncludes /FoCMakeFiles\pangoft2.dir\pango\pango-ot-tag.c.obj /FdCMakeFiles\pangoft2.dir\pangoft2.pdb /FS -c "C:\Program Files (x86)\vcpkg\buildtrees\pango\src\1.40.11-f997ae867d\pango\pango-ot-tag.c"
FAILED: CMakeFiles/pangoft2.dir/pango/pango-ot-tag.c.obj 
C:\PROGRA~2\MICROS~2\2019\COMMUN~1\VC\Tools\MSVC\1422~1.279\bin\Hostx64\x86\cl.exe   -DG_DISABLE_SINGLE_INCLUDES -DG_LOG_DOMAIN=\"Pango\" -DHAVE_CAIRO_FREETYPE=1 -DHAVE_CONFIG_H -DLIBDIR=\"/dummy/lib\" -DPANGO_ENABLE_BACKEND -DPANGO_ENABLE_ENGINE -DSYSCONFDIR=\"/dummy/etc\" -I"C:\Program Files (x86)\vcpkg\buildtrees\pango\src\1.40.11-f997ae867d\." -I"C:\Program Files (x86)\vcpkg\buildtrees\pango\src\1.40.11-f997ae867d\.\pango" -I. -I"C:\Program Files (x86)\vcpkg\installed\x86-windows-static\include" -I"C:\Program Files (x86)\vcpkg\installed\x86-windows-static\include\harfbuzz" /DWIN32 /D_WINDOWS /W3 /utf-8 /MP  /utf-8 /D_DEBUG /MTd /Z7 /Ob0 /Od /RTC1 /showIncludes /FoCMakeFiles\pangoft2.dir\pango\pango-ot-tag.c.obj /FdCMakeFiles\pangoft2.dir\pangoft2.pdb /FS -c "C:\Program Files (x86)\vcpkg\buildtrees\pango\src\1.40.11-f997ae867d\pango\pango-ot-tag.c"
Microsoft (R) C/C++ Optimizing Compiler Version 19.22.27905 for x86
Copyright (C) Microsoft Corporation.  All rights reserved.

C:\Program Files (x86)\vcpkg\buildtrees\pango\src\1.40.11-f997ae867d\pango\pango-ot-private.h(30): fatal error C1083: Cannot open include file: 'hb-glib.h': No such file or directory
ninja: build stopped: subcommand failed.
```

**Comment by bby-co** [comment]
The second issue is resolved by copying `hb-glib.h`, which is residing within `C:\Program Files (x86)\vcpkg\buildtrees\harfbuzz\src\2.5.3-984271955b\src` in my case, to either `C:\Program Files (x86)\vcpkg\packages\harfbuzz_x86-windows-static\include\harfbuzz` or to `C:\Program Files (x86)\vcpkg\installed\x86-windows-static\include\harfbuzz`. I don't know which one fixed it for sure, as I just copied the file into both of them.

So in short, two steps are to be taken to fix this issue:
1-) pango should be declared as dependent on harfbuzz.
2-) On installation of harfbuzz, `hb-glib.h` should be copied to the relevant include directory.

I am leaving this issue open, so that it gets attention.
Feel free to close it after applying the fixes.

**Comment by wangli28** [comment]
Hi @bby-co, thanks for posting this issue!
I can't reproduce this issue locally. I tried to install pango with command ` .\vcpkg.exe install pango:x86-windows-static`, and I got this error when building glib:
```
Starting package 6/15: glib:x86-windows-static
Building package glib[core]:x86-windows-static...
-- Note: glib only supports dynamic library linkage. Building dynamic library.
CMake Error at scripts/cmake/vcpkg_check_linkage.cmake:42 (message):
  Refusing to build unexpected dynamic library against the static CRT.  If
  this is desired, please configure your triplet to directly request this
  configuration.
```
Pango depends on glib and glib doesn't support static library linkage. What did you change before building pango:x86-windows-static?

**Comment by bby-co** [comment]
I have tried building another library statically which have depended on glib, and thus I changed the portfile.cmake to allow static building. This way glib got installed statically.
I wasn't aware that pango was also depending on glib. After making pango "work", the library I tried to originally build was outputting errors because glib didn't support static building, so I just dynamically built what I needed.

So, this was a mistake of mine, sorry.
I am closing the issue now.
